precision testing of knownbits, negative, nonNeg, nonZero, power2, si…

jubitaneja opened the pull request:
…gnbits

jubitaneja commented:
This is WIP, adding test cases and refactoring some parts of code.

regehr commented on include/souper/Extractor/Solver.h:
```diff
@@ -41,6 +41,36 @@ class Solver {
+  virtual
+  std::error_code knownBits(const BlockPCs &BPCs,
+                            const std::vector<InstMapping> &PCs,
+                            Inst *LHS, llvm::APInt &Zeros, llvm::APInt &Ones,
```
let's use llvm::KnownBits instead of a pair of APInts

regehr commented on include/souper/Extractor/Solver.h:
```diff
@@ -41,6 +41,36 @@ class Solver {
+  virtual
+  std::error_code nonNegative(const BlockPCs &BPCs,
+                            const std::vector<InstMapping> &PCs,
+                            Inst *LHS, llvm::APInt &NonNegative,
```
I don't think it's necessary to allocate an APInt for facts like NonNegative, this should just be a bool

regehr commented on include/souper/Extractor/Solver.h:
```diff
@@ -41,6 +41,36 @@ class Solver {
+                            InstContext &IC) = 0;
+  virtual
+  std::error_code powerTwo(const BlockPCs &BPCs,
+                            const std::vector<InstMapping> &PCs,
```
careful with indentation here and elsewhere -- make sure your browser has no tabs

regehr commented on lib/Extractor/Candidates.cpp:
```diff
@@ -847,6 +871,64 @@ void ExtractExprCandidates(Function &F, const LoopInfo *LI, DemandedBits *DB,
+        if (Negative)
+          llvm::outs() << "known at return: " << "(negative)" << "\n";
+        else
+          llvm::outs() << "known at return: " << "" << "\n";
```
how about just print something like: "known negative?" followed by true or false? otherwise all these blank "known at return:" are hard to interpret

regehr commented on lib/Extractor/Candidates.cpp:
```diff
@@ -847,6 +871,64 @@ void ExtractExprCandidates(Function &F, const LoopInfo *LI, DemandedBits *DB,
+      if (PrintNonNegAtReturn && isa<ReturnInst>(I)) {
+        auto V = I.getOperand(0);
+        auto DL = F.getParent()->getDataLayout();
+        bool NonNegative = 0;
```
false, not 0

regehr commented on lib/Extractor/Candidates.cpp:
```diff
@@ -847,6 +871,64 @@ void ExtractExprCandidates(Function &F, const LoopInfo *LI, DemandedBits *DB,
+        auto V = I.getOperand(0);
+        auto DL = F.getParent()->getDataLayout();
+        bool NonNegative = 0;
+        NonNegative = isKnownNonNegative(V, DL);
```
but also no initial value needed if you're just going to overwrite it

regehr commented on lib/Extractor/Solver.cpp:
```diff
@@ -76,6 +76,164 @@ class BaseSolver : public Solver {
+                           InstContext &IC) override {
+    unsigned W = LHS->Width;
+    Negative = APInt::getNullValue(W);
+    APInt NegativeGuess = Negative | APInt::getOneBitSet(W, W-1);
```
why do you want to do math using a variable that you just set to zero?

regehr commented on lib/Extractor/Solver.cpp:
```diff
@@ -76,6 +76,164 @@ class BaseSolver : public Solver {
+    } else if (testOneSign(BPCs, PCs, NonNegativeGuess, LHS, IC)) {
+      NonNegative = NonNegativeGuess;
+    } else {
+      NonNegative = NonNegativeGuess;
```
it looks like this else branch and the previous "else if" branch contain the same code, that cannot be correct

regehr commented on lib/Extractor/Solver.cpp:
```diff
@@ -76,6 +76,164 @@ class BaseSolver : public Solver {
+
+      if (!IsSat) {
+        SignBits = I;
+        continue;
```
continue not needed

regehr commented on CMakeLists.txt:
```diff
@@ -409,6 +409,12 @@ set(SOUPER_PASS_PROFILE_ALL ${CMAKE_BINARY_DIR}/${CMAKE_SHARED_LIBRARY_PREFIX}so
+configure_file(${CMAKE_SOURCE_DIR}/utils/cache_nonNeg.in ${CMAKE_BINARY_DIR}/cache_nonNeg @ONLY)
+configure_file(${CMAKE_SOURCE_DIR}/utils/cache_power.in ${CMAKE_BINARY_DIR}/cache_power @ONLY)
+configure_file(${CMAKE_SOURCE_DIR}/utils/cache_nonZero.in ${CMAKE_BINARY_DIR}/cache_nonZero @ONLY)
+configure_file(${CMAKE_SOURCE_DIR}/utils/cache_signbits.in ${CMAKE_BINARY_DIR}/cache_signbits @ONLY)
```
it seems like these are all very similar scripts, can't they be collapsed into one?

jubitaneja commented:
Thanks for the review comments. I will fix it soon.

Jubi

On Tue, Apr 16, 2019, 9:10 PM John Regehr <notifications@github.com> wrote:

> *@regehr* commented on this pull request.
> ------------------------------
>
> In CMakeLists.txt
> <https://github.com/google/souper/pull/536#discussion_r276072745>:
>
> > @@ -409,6 +409,12 @@ set(SOUPER_PASS_PROFILE_ALL ${CMAKE_BINARY_DIR}/${CMAKE_SHARED_LIBRARY_PREFIX}so
>  set(PROFILE_LIBRARY ${CMAKE_BINARY_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}profileRuntime${CMAKE_STATIC_LIBRARY_SUFFIX})
>  configure_file(${CMAKE_SOURCE_DIR}/utils/reduce.in ${CMAKE_BINARY_DIR}/reduce @ONLY)
>  configure_file(${CMAKE_SOURCE_DIR}/utils/cache_dump.in ${CMAKE_BINARY_DIR}/cache_dump @ONLY)
> +configure_file(${CMAKE_SOURCE_DIR}/utils/cache_knownbits.in ${CMAKE_BINARY_DIR}/cache_knownbits @ONLY)
> +configure_file(${CMAKE_SOURCE_DIR}/utils/cache_neg.in ${CMAKE_BINARY_DIR}/cache_neg @ONLY)
> +configure_file(${CMAKE_SOURCE_DIR}/utils/cache_nonNeg.in ${CMAKE_BINARY_DIR}/cache_nonNeg @ONLY)
> +configure_file(${CMAKE_SOURCE_DIR}/utils/cache_power.in ${CMAKE_BINARY_DIR}/cache_power @ONLY)
> +configure_file(${CMAKE_SOURCE_DIR}/utils/cache_nonZero.in ${CMAKE_BINARY_DIR}/cache_nonZero @ONLY)
> +configure_file(${CMAKE_SOURCE_DIR}/utils/cache_signbits.in ${CMAKE_BINARY_DIR}/cache_signbits @ONLY)
>
> it seems like these are all very similar scripts, can't they be collapsed
> into one?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/souper/pull/536#pullrequestreview-227542645>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGS3lLrKBSF1dfJtET49J1A-1o4zBLDRks5vhp6ogaJpZM4cw3_j>
> .
>

jubitaneja commented:
I have fixed and updated this PR. souper2llvm in this PR has changes to support dataflow facts only. About creating one big script for testing dataflow facts -- we can do that, but this will be a bit complicated to read and debug if required later. Maybe keeping individual scripts is an easy way to use it (just in my opinion). If you think we should really have just one perl script, i can spend some time putting it all together.

regehr commented:
I thought you were going to move the souper2llvm changes into a different pull request

regehr commented:
also I don't want a bunch of different scripts for inferring dataflow facts from a cache, you need to merge these. this code is impossible to maintain

jubitaneja commented:
I kept the minimal souper2llvm changes in this PR that only takes care of translating dataflow facts in Souper IR to LLVM IR. I will take away every change in that case and put it all in next PR. Also, that would mean we cannot test dataflow facts on LLVM side using commands like: 
` %souper2llvm %s | %llvm-as | %souper %solver -print-known-at-return`

jubitaneja commented:
Sure, will merge all scripts of caching into one.

regehr commented on lib/Extractor/Candidates.cpp:
```diff
@@ -847,6 +871,64 @@ void ExtractExprCandidates(Function &F, const LoopInfo *LI, DemandedBits *DB,
+      if (PrintNegAtReturn && isa<ReturnInst>(I)) {
+        auto V = I.getOperand(0);
+        auto DL = F.getParent()->getDataLayout();
+        bool Negative;
```
in this case and others, you should prefer this style:
```bool Negative = isKnownNegative(V, DL);```

regehr replied:
of course "auto" would be fine too

regehr commented on lib/Extractor/Candidates.cpp:
```diff
@@ -847,6 +871,64 @@ void ExtractExprCandidates(Function &F, const LoopInfo *LI, DemandedBits *DB,
+        else
+          llvm::outs() << "known powerOfTwo from compiler: " << "false" << "\n";
+      }
+      if (PrintNonZeroAtReturn && isa<ReturnInst>(I)) {
```
instead of these repeated tests of ```isa<ReturnInst>``` just test this once at the beginning and then test the individual flags

regehr commented on lib/Extractor/Solver.cpp:
```diff
@@ -76,6 +76,157 @@ class BaseSolver : public Solver {
       : SMTSolver(std::move(SMTSolver)), Timeout(Timeout) {}
 
+  bool testZeroMSB(const BlockPCs &BPCs,
+                    const std::vector<InstMapping> &PCs,
```
indentation problem

regehr commented on lib/Extractor/Solver.cpp:
```diff
@@ -76,6 +76,157 @@ class BaseSolver : public Solver {
+                           InstContext &IC) override {
+    unsigned W = LHS->Width;
+    Negative = false;
+    if (testZeroMSB(BPCs, PCs, LHS, IC))
```
I don't understand this, why are you calling a function and then using its result to set Negative to false, when you just initialized Negative to false one line earlier in the code?

regehr commented on lib/Extractor/Solver.cpp:
```diff
@@ -76,6 +76,157 @@ class BaseSolver : public Solver {
+    NonNegative = false;
+    if (testZeroMSB(BPCs, PCs, LHS, IC)) {
+      NonNegative = true;
+    } else if (testOneMSB(BPCs, PCs, LHS, IC)) {
```
same here

regehr commented on lib/Extractor/Solver.cpp:
```diff
@@ -76,6 +76,157 @@ class BaseSolver : public Solver {
+
+  std::error_code powerTwo(const BlockPCs &BPCs,
+                           const std::vector<InstMapping> &PCs,
+                           Inst *LHS, APInt &PowTwo,
```
PowTwo should be a bool, not an APInt

regehr commented on lib/Extractor/Solver.cpp:
```diff
@@ -76,6 +76,157 @@ class BaseSolver : public Solver {
+    std::error_code EC = SMTSolver->isSatisfiable(BuildQuery(IC, BPCs, PCs, Mapping, 0, /*Precondition=*/0),
+                                                  IsSat, 0, 0, Timeout);
+    if (EC)
+      llvm::report_fatal_error("stopping due to error");
```
in this case and others, say something useful when an error occurs, ```stopping due to error``` is not useful to someone running this code

regehr commented on lib/Extractor/Solver.cpp:
```diff
@@ -76,6 +76,157 @@ class BaseSolver : public Solver {
+
+  std::error_code nonZero(const BlockPCs &BPCs,
+                          const std::vector<InstMapping> &PCs,
+                          Inst *LHS, APInt &NonZero,
```
bool

regehr commented on test/Infer/precision-kb-2.opt:
```diff
@@ -0,0 +1,11 @@
+; RUN: %souper2llvm %s | %llvm-as | %souper %solver -print-known-at-return
+
+; CHECK: knownBits from souper: xxxxxxx0
+; CHECK: knownBits from compiler: xxxxxxxx
```
it's great to test that souper returns the precise result, but do not test that the compiler returns the imprecise result, that is not something that we want our tests relying upon

jubitaneja replied:
I added this check to only test what LLVM returns, but not with an intent to compare the precision of Souper vs LLVM. But, I will take off this LLVM check part from here and add relevant tests for LLVM where we can expect some useful dataflow information.

regehr commented on test/Infer/precision-neg-1.opt:
```diff
@@ -0,0 +1,12 @@
+; RUN: %souper-check %solver -infer-rhs -infer-neg %s
+; RUN: %souper2llvm %s | %llvm-as | %souper %solver -print-neg-at-return
+
+; CHECL: known negative from Souper:   true
```
mis-spelled word CHECL

regehr commented on test/Infer/precision-neg-2.opt:
```diff
@@ -0,0 +1,12 @@
+; RUN: %souper-check %solver -infer-rhs -infer-neg %s
+; RUN: %souper2llvm %s | %llvm-as | %souper %solver -print-neg-at-return
+
+; CHECL: known negative from Souper:   true
```
same

regehr commented on test/Infer/precision-nonNeg-1.opt:
```diff
@@ -0,0 +1,17 @@
+; RUN: %souper-check %solver -infer-rhs -infer-non-neg %s
+; RUN: %souper2llvm %s | %llvm-as | %souper %solver -print-nonneg-at-return
+
+; CHECL: known nonNegative from Souper:   true
```
same

regehr commented on test/Infer/precision-nonNeg-2.opt:
```diff
@@ -0,0 +1,13 @@
+; RUN: %souper-check %solver -infer-rhs -infer-non-neg %s
+; RUN: %souper2llvm %s | %llvm-as | %souper %solver -print-nonneg-at-return
+
+; CHECL: known nonNegative from Souper:   true
```
mis-spelled word CHECL

regehr commented on test/Infer/precision-nonNeg-3.opt:
```diff
@@ -0,0 +1,11 @@
+; REQUIRES: solver
+; RUN: %souper-check %solver -infer-rhs -infer-non-neg %s
+
+; CHECL: known nonNegative from Souper:   true
```
mis-spelled word CHECL

regehr commented on tools/souper-check.cpp:
```diff
@@ -96,6 +121,93 @@ int SolveInst(const MemoryBufferRef &MB, Solver *S) {
+      if (Negative)
+        s = "true";
+      else
+        s = "false";
```
avoid this repeated code for making a string out of a bool. either factor it out into a helper function or else use the native C++ method for printing a bool as a string.

jubitaneja replied:
std::boolalpha didn't work in context of llvm::outs(), so I added a small helper function.

jubitaneja commented:
Thanks for the detailed feedback. Updated the PR again.

regehr commented on test/Infer/precision-kb-5.opt:
```diff
@@ -0,0 +1,19 @@
+
+
+; CHECK: knownBits from souper: 00000000000000xxxxxxxxxxxxxxxxxx
+; CHECK: knownBits from compiler: 0000000000000xxxxxxxxxxxxxxxxxxx
```
don't test what the compiler returns

regehr commented on test/Infer/precision-neg-1.opt:
```diff
@@ -0,0 +1,12 @@
+; RUN: %souper2llvm %s | %llvm-as | %souper %solver -print-neg-at-return
+
+; CHECK: known negative from Souper:   true
+; CHECK: known negative from compiler: false
```
don't test what the compiler returns

regehr commented on test/Infer/precision-power-3.opt:
```diff
@@ -0,0 +1,12 @@
+; RUN: %souper2llvm %s | %llvm-as | %souper %solver -print-power-two-at-return
+
+; CHECK: known powerOfTwo from souper:   true
+; CHECK: known powerOfTwo from compiler: false
```
fix all of these

regehr commented on tools/souper-check.cpp:
```diff
@@ -96,6 +128,74 @@ int SolveInst(const MemoryBufferRef &MB, Solver *S) {
+      bool NonZero;
+      if (std::error_code EC = S->nonZero(Rep.BPCs, Rep.PCs, Rep.Mapping.LHS,
+                                              NonZero, IC)) {
+        llvm::errs() << EC.message() << '\n';
```
in all of these functions you must check the error code, not just print it out. the dataflow result is not valid when there's an error, so the code can't just keep going.

jubitaneja replied:
Fixed.

regehr commented:
LGTM, thanks!

LebedevRI commented:
> I kept the minimal souper2llvm changes in this PR that only takes care of translating dataflow facts in Souper IR to LLVM IR. I will take away every change in that case and put it all in next PR. Also, that would mean we cannot test dataflow facts on LLVM side using commands like:
> ` %souper2llvm %s | %llvm-as | %souper %solver -print-known-at-return`

I keep oscillating between writing too much words, and not enough words it seems.
I was suggesting making the `souper2llvm.in` fixes *FIRST*, because currently that tool is still broken.
Really looking forward to the next PR with those changes i guess..

(Related: is static optimization profile working?
`cache_dump` tells me it is always zero, even though `SOUPER_STATIC_PROFILE` is set)

jubitaneja commented:
@LebedevRI Thanks for checking in on souper2llvm status. @zhengyangl and I are working on it, will soon send a PR for review. In case, you want to have a look at current souper2llvm fixes only, you can check this branch: https://github.com/jubitaneja/souper/tree/fix-souper2llvm

LebedevRI commented:
Thanks!

> About static profile, I will have to check again. I remember using this script a couple of months ago and it worked fine.

Yes, that is my recollection, too..
